tree-walk-pool.go

```diff
@@ -77,60 +87,67 @@ func (t treeWalkerPool) Release(params listParams) (treeWalkerCh chan treeWalker
 			} else {
 				delete(t.pool, params)
 			}
-			walk.doneCh <- struct{}{}
-			return walk.treeWalkerCh, walk.treeWalkerDoneCh
+			walk.endTimerCh <- struct{}{}
+			return walk.resultCh, walk.endWalkCh
 		}
 	}
 	// Release return nil if params not found.
 	return nil, nil
 }
 
-// Set - adds new list params along with treeWalker channel to the
-// pool for future. Additionally this also starts a go routine which
-// waits at the configured timeout. Additionally this go-routine is
-// also closed pro-actively by 'Release' call when the treeWalker
-// item is obtained from the pool.
-func (t treeWalkerPool) Set(params listParams, treeWalkerCh chan treeWalker, treeWalkerDoneCh chan struct{}) {
+// Set - adds a treeWalk to the treeWalkPool.
+// Also starts a timer go-routine that ends when:
+// 1) time.After() expires after t.timeOut seconds.
+//    The expiration is needed so that the treeWalk go-routine resources are freed after a timeout
+//    if the S3 client does only partial listing of objects.
+// 2) Relase() signals the timer go-routine to end on endTimerCh.
+//    During listing the timer should not timeout and end the treeWalk go-routine, hence the
+//    timer go-routine should be ended.
+func (t treeWalkPool) Set(params listParams, resultCh chan treeWalkResult, endWalkCh chan struct{}) {
 	t.lock.Lock()
 	defer t.lock.Unlock()
 
 	// Should be a buffered channel so that Release() never blocks.
-	var doneCh = make(chan struct{}, 1)
-	walkInfo := treeWalkerPoolInfo{
-		treeWalkerCh:     treeWalkerCh,
-		treeWalkerDoneCh: treeWalkerDoneCh,
-		doneCh:           doneCh,
+	endTimerCh := make(chan struct{}, 1)
+	walkInfo := treeWalk{
+		resultCh:   resultCh,
+		endWalkCh:  endWalkCh,
+		endTimerCh: endTimerCh,
 	}
 	// Append new walk info.
 	t.pool[params] = append(t.pool[params], walkInfo)
 
-	// Safe expiry of treeWalkerCh after timeout.
-	go func(doneCh <-chan struct{}) {
+	// Timer go-routine which times out after t.timeOut seconds.
```

Comment:
We should not have to write obvious things saying that 'go func()' is a 'go-routine' is superfluous.  Previous comment is better it talks about what this go routine does.